src/libmongoc/src/mongoc/mongoc-ts-pool-private.h

```diff
@@ -0,0 +1,97 @@
+#ifndef MONGOC_TS_POOL_PRIVATE_H
+#define MONGOC_TS_POOL_PRIVATE_H
+
+#include <stddef.h>
+#include <stdbool.h>
+
+/**
+ * @brief A thread-safe pool.
+ *
+ * Items are allocated using `mongo_ts_pool_alloc_item` and should be deleted
+ * with `mongo_ts_pool_free_item`. Items are added to the pool with
+ * `mongo_ts_pool_push` and taken from the pool using `mongo_ts_pool_try_pop`.
+ */
+typedef struct _mongoc_ts_pool *mongoc_ts_pool;
+
+/**
+ * @brief Create a new thread-safe pool
```

Comment:
Awesome, this seems generally useful!

Can you add unit tests of `mongoc_ts_pool` (akin to test-mongoc-generation-map.c)?

If we need to modify this data structure in the future, I think it would be easier to identify regressions in unit tests.